src/Passes/Source/Commandline/ConfigurationManager.hpp

```diff
@@ -0,0 +1,217 @@
+        void printConfiguration() const;
+
+        // Configuration functions
+        //
+
+        /// Adds a new configuration of type T.
+        template <typename T> inline void addConfig(String const& id = "", T const& default_value = T());
+
+        /// Whether or not the component associated with T is active.
+        template <typename T> inline bool isActive();
+
+        /// Sets the section name. This method is used by the configuration class to set a section
+        /// name.
+        void setSectionName(String const& name, String const& description);
+
+        /// Adds a new parameter with a default value to the configuration section. This function should
+        /// be used by the configuration class.
+        template <typename T>
+        inline void addParameter(T& bind, T default_value, String const& name, String const& description);
+
+        /// Adds a new parameter to the configuration section. This method uses the bound variable value
+        /// as default value. This function should be used by the configuration class.
+        template <typename T> inline void addParameter(T& bind, String const& name, String const& description);
+
+      private:
+        Sections config_sections_{}; ///< All available sections within the ConfigurationManager instance
+    };
+
+    template <typename T> inline void ConfigurationManager::addConfig(String const& id, T const& default_value)
+    {
+        Section new_section{std::type_index(typeid(T))};
+
+        auto ptr                  = std::make_shared<T>(default_value);
+        new_section.configuration = ptr;
+        new_section.active        = std::make_shared<bool>(true);
+        new_section.id            = id;
+
+        config_sections_.emplace_back(std::move(new_section));
+        ptr->setup(*this);
+    }
+
+    template <typename T> inline void ConfigurationManager::setConfig(T const& value)
+    {
+        auto    type = std::type_index(typeid(T));
+        VoidPtr ptr{nullptr};
+
+        for (auto& section : config_sections_)
+        {
+            if (section.type == type)
+            {
+                ptr = section.configuration;
+                break;
+            }
+        }
+
+        if (ptr == nullptr)
+        {
+            throw std::runtime_error("Could not find configuration class.");
+        }
+
+        auto& config = *static_cast<T*>(ptr.get());
+        config       = value;
+    }
+
+    template <typename T> inline T const& ConfigurationManager::get() const
+    {
+        VoidPtr ptr{nullptr};
+        auto    type = std::type_index(typeid(T));
+
+        for (auto& section : config_sections_)
+        {
+            if (section.type == type)
+            {
+                ptr = section.configuration;
+                break;
+            }
+        }
+
+        if (ptr == nullptr)
+        {
+            throw std::runtime_error("Could not find configuration class.");
+        }
+
+        return *static_cast<T*>(ptr.get());
+    }
```

Comment:
Code is mostly the same between the former (`ConfigurationManager::setConfig()`) and the latter (`ConfigurationManager::get()`). Consider implementing the former by calling the latter (and the latter returning a reference to non-const).

Reply:
Yes, you are right. Technically your suggestion is not possible as `get` returns a const reference which would not allow setting the value. Instead, I've created an internal helper function and moved most of the get implementation there.